### PR TITLE
"renewDeadline" support for leader election

### LINKED ISSUE #706: "renewDeadline" support for leader election
https://github.com/kubernetes-client/java/blob/5ae91646b465774c194a1b9299274005c185d5d6/extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElectionConfig.java#L11

The "renewDeadline" field in `LeaderElectionConfig` seems to be useless currently, is there a plan to support renew deadline duration in the future?
Probably related to this https://github.com/kubernetes-client/java/pull/538#discussion_r274320762

### PR DESCRIPTION
**Pull request by moeyui1**
Fix #706
this pr fix the same issue as #708 , cause I think #708 got some misunderstanding with "renewDeadline".

### CODE REVIEW

**Comment by moeyui1** [comment]
I signed it

File: extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java [hunk 1]
```diff
@@ -96,34 +132,43 @@ private void renewLoop() {
...
+                    }
+                  } catch (InterruptedException e) {
+                    return false;
+                  }
+                  return true;
+                });
+
+        boolean renewResult;
+        try {
+          renewResult = future.get(renewDeadlineMillis, TimeUnit.MILLISECONDS);
```
Comment (yue9944882) [inline extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java]: applying timeout here is a good way to mock out the channel-based timer's behavior in golang i agree.

**Comment by yue9944882** [comment]
cc @wu8685 for thoughts

**Comment by brendandburns** [comment]
This looks good to me, I defer to others for final review before submitting...

/approve

File: extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java [hunk 2]
```diff
@@ -96,34 +132,43 @@ private void renewLoop() {
...
+                    while (!tryAcquireOrRenew()) {
+                      Thread.sleep(retryPeriodMillis);
+                    }
+                  } catch (InterruptedException e) {
+                    return false;
+                  }
+                  return true;
+                });
+
+        boolean renewResult;
```
[the remark this answers is not in the document]
Comment (moeyui1) [reply extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java]: may be something like this?
```java
      while (true) {
        final Future<Boolean> future =
            leaseWorkers.submit(
                () -> {
                  try {
                    // retry until success or interrupted
                    while (!tryAcquireOrRenew()) {
                      Thread.sleep(retryPeriodMillis);
                    }
                  } catch (InterruptedException e) {
                    return false;
                  }
                  return true;
                });
        try {
          if (future.get(renewDeadlineMillis, TimeUnit.MILLISECONDS)) {
            if (log.isDebugEnabled()) {
              log.debug("Successfully renewed lease");
            }
          } else {
            break;
          }
        } catch (ExecutionException | TimeoutException t) {
          break;
        } finally {
          future.cancel(true);
        }
      }
```
It's OK but I prefer to keep the try-catch simple and clear. @brendanburns @yue9944882 What do you think?

**Comment by moeyui1** [comment]
Is there anything else I should modify? Thought it's ready to be merged.

**Comment by yue9944882** [comment]
> it needs a unit test for verification, otherwise LGTM

please add a unit test to verify out the behavior

**Comment by moeyui1** [comment]
> > it needs a unit test for verification, otherwise LGTM
> 
> please add a unit test to verify out the behavior

Sure, I will add some tests soon.

**Comment by yue9944882** [comment]
thanks for fixing this!

/lgtm

**Pull request merged**